### PR TITLE
add sharded adam

### PR DESCRIPTION
**Pull request by ver217**

### CODE REVIEW

File: colossalai/zero/sharded_optim/sharded_adam.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        self.grad_scaler = DynamicGradScaler(initial_scale=initial_scale,
+                                             min_scale=min_scale,
+                                             growth_factor=growth_factor,
+                                             backoff_factor=backoff_factor,
+                                             growth_interval=growth_interval,
+                                             hysteresis=hysteresis,
+                                             max_scale=max_scale)
+        self._found_overflow: Tensor = torch.FloatTensor([0]).to(self.state_device)
+
+        # Early state initialization
```
Comment (ver217) [inline colossalai/zero/sharded_optim/sharded_adam.py:56]: @feifeibear I think the sharded Adam is easy to expand toa  general sharded optimizer by just removing early state initialization here. And in step(), we have to make sure param.data has correct shape, dtype and device.

File: colossalai/zero/sharded_optim/sharded_adam.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+        found_inf = self._check_overflow()
+        self.grad_scaler.update(found_inf)
+
+        if found_inf:
+            self.zero_grad()
+            return
+
+        # Write payload back to p.data
+        for group in self.optim.param_groups:
```
Comment (ver217) [inline colossalai/zero/sharded_optim/sharded_adam.py:97]: @feifeibear I write payload back to param.data here. Also, this function can handle unsharded (normal) parameters as well.

File: colossalai/zero/sharded_optim/sharded_adam.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                if hasattr(p, 'ca_attr'):
+                    data = p.ca_attr.payload(self.state_device)
+                if torch.is_floating_point(data) and data.dtype != torch.float:
+                    data = data.to(torch.float)
+                p.data = data
+        ret = self.optim.step(*args, **kwargs)
+        # Set p.data to None
+        for group in self.optim.param_groups:
+            for p in group['params']:
+                p.data = None
```
Comment (ver217) [inline colossalai/zero/sharded_optim/sharded_adam.py:109]: @feifeibear I should write the updated weight to sharded param as well. Do we have a method to update it?

File: colossalai/zero/sharded_model/sharded_model_v2.py [hunk 4]
```diff
@@ -81,6 +83,10 @@ def backward(self, loss):
         loss.backward()
         self._final_backward_hook()
 
+    def backward_by_grad(self, tensor, grad):
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_model_v2.py:86]: Why we have is API? Did PyTorch nn.Module has one?

File: colossalai/zero/sharded_optim/sharded_adam.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+class OptimState(Enum):
+    SCALED = 1
+    UNSCALED = 2
+
+
+class ShardedAdam(ColossalaiOptimizer):
+
+    def __init__(self,
+                 adam_optim: Optimizer,
+                 sharded_model: nn.Module,
```
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_adam.py:28]: nn.Module -> Union[nn.Module, ShardedModelV2]

File: colossalai/zero/sharded_optim/sharded_adam.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_adam.py:56]: I agree with you on moving OS memory allocation and init here.

**Comment by FrankLeeeee** [comment]
Hi, sharded_model_V2 has conflict.

File: colossalai/zero/sharded_optim/sharded_adam.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_adam.py:97]: I have discussed with ver217. The following code snippet is wrong.

File: colossalai/zero/sharded_optim/sharded_adam.py (hunk 1 above)
Comment (FrankLeeeee) [reply colossalai/zero/sharded_optim/sharded_adam.py:56]: Agree. This part is not necessary as many torch optimizers will do lazy init.

File: colossalai/zero/sharded_model/sharded_model_v2.py (hunk 4 above)
Comment (FrankLeeeee) [reply colossalai/zero/sharded_model/sharded_model_v2.py:86]: This API is in fact for pipeline parallel. The pipeline stage which is not the last stage has to backward with grad and tensor.

**Review by feifeibear — APPROVED** [review]
LGTM.
I will add grad fp16->fp32 and param fp32->fp16 copy in another PR.

**Comment by ver217** [comment]
> 

In sharded_grad, I store a fp32 copy after reduction. So the grad is fp32 when calling `optim.step()`.

**Pull request merged**